### PR TITLE
feat(codegen): Detect repeated struct field when deserializing

### PR DESCRIPTION
**Pull request by dtolnay**
Addresses #59. Let me know whether you think we need an escape hatch to opt out of this check.

### CODE REVIEW

**Comment by oli-obk** [comment]
> Let me know whether you think we need an escape hatch to opt out of this check.

I don't think we need one, but technically this is a breaking change. I wonder if anyone relies on it. Maybe toml?

**Comment by dtolnay** [comment]
@alexcrichton any idea whether the TOML serde relies on the behavior that the last value for a particular struct field overrides any previous values for the same struct field? Using JSON syntax, this PR proposes that deserializing `{"a": 1, "a": 2}` should be an error rather than `MyStruct { a: 2 }`.

**Comment by erickt** [comment]
cc @alexcrichton on the toml question. Would this cause you problems?

**Comment by alexcrichton** [comment]
Ah yeah I don't think that this is a problem for TOML, the format already disallows duplicate keys in a table, so it's actually impossible to have a TOML document with two key/value pairs that have the same key.

So sounds good to me!

**Comment by dtolnay** [comment]
I have resolved the merge conflict with #285.

**Comment by dtolnay** [comment]
Because of the `use ::serde::de::Error` there is a semantic merge conflict with #298 so one of these PRs will break the other. It looks like @erickt has reviewed #298 already so let's merge that first and then @erickt or I can rebase this one.

**Comment by dtolnay** [comment]
I am debugging why this does not compile after the merge commit.

**Comment by dtolnay** [comment]
I narrowed this down to the following difference. This compiles:

```rust
extern crate serde;
use serde::de::Error;
fn main() {}
```

This does not compile:

```rust
fn main() {
    extern crate serde;
    use serde::de::Error;
}
```

The compiler diagnostic says "Did you mean \`self::serde::de\`?"

If I change it to `self::serde::de`, it still fails to compile and the diagnostic says "Did you mean \`self::self::serde::de\`?"

If I change it to `self::self::serde::de` (which is nonsensical), the diagnostic says "Maybe a missing \`extern crate self\`?"

Has anyone seen this before or know a workaround?

**Comment by dtolnay** [comment]
Resolved with the help of [this thread](https://users.rust-lang.org/t/use-trait-from-crate-in-the-same-lexical-scope/5671) on the users forum. The remaining failure is related to [quasi](https://travis-ci.org/serde-rs/serde/jobs/127422523) not this PR. Ready for another review @oli-obk or @erickt.

File: serde_tests/tests/token.rs [hunk 1]
```diff
@@ -429,6 +429,10 @@ impl de::Error for Error {
 
     fn end_of_stream() -> Error { Error::EndOfStreamError }
 
+    fn invalid_value(msg: &str) -> Error {
+        Error::InvalidValue(msg.to_owned())
```
Comment (oli-obk) [inline serde_tests/tests/token.rs]: I think `de::Error` deserves a new method `duplicate_field`. We already have `unknown_field`. The default impl could forward to `invalid_value` for now, and we can add an issue for 0.8 to remove that default impl.

Comment (dtolnay) [inline serde_tests/tests/token.rs]: Sounds reasonable. I pushed the change. Why do you say we need to get rid of the default impl in 0.8?

Comment (oli-obk) [inline serde_tests/tests/token.rs]: Because I was confused on how Error is implemented ;)

**Comment by dtolnay** [comment]
The Travis failure is due to clippy.

**Comment by dtolnay** [comment]
I pushed a merge commit to fix the conflict with #308.

**Comment by dtolnay** [comment]
The Travis failure is due to #313.

**Comment by dtolnay** [comment]
Rebased, per https://github.com/serde-rs/serde/pull/315#issuecomment-218154729.

**Comment by oli-obk** [comment]
@homu r+

**Pull request merged**